### DISCUSSION TITLE
Lack of documentation on Minio reliability

### DISCUSSION THREAD

**Discussion by gytis-ivaskevicius** [minio/minio#13826] [Q&A] [answered]
Hi,
I am interested in how minio handles data, I've found some information [here](https://docs.min.io/docs/distributed-minio-quickstart-guide.html) but not enough.

Here are a couple of questions, some links or answers, or even code references would be appreciated.
- Does minio persist objects to all nodes at the same time?
- When 'putObject' is considered successful
  -  When at least one node received all the data?
  -  When at least one node persisted all the data?
  -  When multiple nodes receive all the data?
  -  When multiple nodes persisted all the data?
  - Something else?
- Can minio act as raid0? 
- Can minio persist the same bucket on multiple nodes without acting as raid1?
- Overall, is there a concept of sharding?
- How many nodes can go down? (50%?)
- If only some of the nodes contain requested objects - but node requested does not have it - does it internally request node having that data and forwards it? (Even if there is no sharding such case can occur when node is starting up, or maybe it does not accept any requests until data gets synced?)

Also maybe it would be nice to put answers to https://docs.min.io/ ?

**Comment by klauspost** [comment, marked as answer]
I will not repeat information that is available. For example here: https://docs.min.io/minio/baremetal/concepts/erasure-coding.html#erasure-sets

> Does minio persist objects to all nodes at the same time?

All drives in a set that the contains the object are written to concurrently.

> When 'putObject' is considered successful

When enough disks have successfully written the data for erasure code restruction to be successful. If drives are offline when a write starts, we add additional erasure codes to compensate.

If we cannot reach that goal, we don't keep the object and return an error.

> When at least one node received all the data?

Nodes have no special meaning.

> When at least one node persisted all the data?

It cannot. Data is sharded across nodes.

> Can minio act as raid0?

In concept it is closest to raid6, but with more erasure (parity) blocks, up to 8 of 16 drives.

> Can minio persist the same bucket on multiple nodes without acting as raid1?

Yes. MinIO has several [replication options](https://docs.min.io/minio/baremetal/replication/replication-overview.html)

> Overall, is there a concept of sharding?

Yes. All objects are sharded.

> How many nodes can go down? (50%?)

That depends on your number of servers and disks: Use the [Erasure Code Calculator](https://min.io/product/erasure-code-calculator?ref=docs)

>If only some of the nodes contain requested objects - but node requested does not have it - does it internally request node having that data and forwards it?

Objects are placed deterministically within a server pool based on their key. Only if you have several server pools (via expansion) will it need to ask each server pool where an object is located.

TBH it doesn't really seem like you read the documentation available. Most of this information is already available at https://docs.min.io/minio/baremetal/

**Reply by ravindk89** [reply]
I'll note we are working on replacing the root docs.min.io entirely to make it easier to find the newer docs. Our main sites list the Baremetal docs first as part of the Docs dropdown. What is left now is fixing things so Google points to the newest available information.
